### PR TITLE
[WIP] MobileNet model

### PR DESCRIPTION
**Pull request by Maratyszcza**
**DO NOT MERGE. WORK IN PROGRESS**

Implementation of the model from the [MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications](https://arxiv.org/abs/1704.04861) paper.

### CODE REVIEW

**Comment by revilokeb** [comment]
@Maratyszcza nice, thanks! I have seen in the PR that pretrained weights (presumably based on ILSVRC12) will be available, too. Can I ask about the timeline for this? Else I might consider training the architecture myself...

**Comment by Maratyszcza** [comment]
I'm training model with two sets of hyperparameters now. Should finish in a week, assuming it converges to reported accuracy.

**Comment by revilokeb** [comment]
@Maratyszcza Can I ask if you are observing good convergence? Training on my side is progressing but fairly slowly. Generally it seems to me that grouped convs currently are maybe more efficient in TF (https://github.com/Zehaos/MobileNet vs this or https://github.com/marvis/pytorch-mobilenet)

**Comment by Maratyszcza** [comment]
Training with learning rate `0.1` converged to 66% Top-1, quite a bit less than the paper. My machine is now busy with other things, but I'll retrain with learning rate `0.01` when its available.

cuDNN v6 doesn't support grouped convolutions, so PyTorch calls cuDNN once per group. As MobileNets have up to 1024 groups in conv layer, thats a lot of kernel launches. TF has a special implementation for depthwise convolution, which doesn't use cuDNN. Similar implementation for PyTorch was merged just hours ago: https://github.com/torch/nn/pull/1098

**Comment by revilokeb** [comment]
Good piece of info, thanks!

**Comment by eladhoffer** [comment]
I managed to train it to ~69% Top-1 accuracy by following the resnet training regime.

**Comment by jtoy** [comment]
@Maratyszcza that merge is for regular torch, not pytorch  does that mean pytorch will get depthwise convolution?

**Comment by Maratyszcza** [comment]
@jtoy libTHNN is shared between Lua Torch and PyTorch.

**Comment by illarion-rl** [comment]
Hi!
Can someone advice what is the progress of this implementation?

**Comment by Maratyszcza** [comment]
@illarion-rl I didn't manage to get the accuracy reported in the paper and eventually gave up. Feel free to take over.

**Comment by fmassa** [comment]
@Coderx7 isn't [`RandomResizedCrop`](https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py#L519) the pre-processing that Inception uses?

**Comment by Coderx7** [comment]
@fmassa apparently no  based on [this](https://discuss.pytorch.org/t/how-do-i-replicate-the-exact-preprocessing-for-imagenet-2012-validation/13899) and [this](https://github.com/pytorch/vision/pull/429)

**Comment by eladhoffer** [comment]
Using this version I was able to reach 70%:
https://github.com/eladhoffer/convNet.pytorch/blob/master/models/mobilenet.py

**Comment by Coderx7** [comment]
@eladhoffer : Thanks a lot. 
I'm a bit puzzled with how you trained your model!
ther preprocessing seems normal, however, whats with all of those regimes and different input_sizes? : 
```
        if regime == 'small':
            scale_lr = 4
            self.regime = [
                {'epoch': 0, 'optimizer': 'SGD',
                    'momentum': 0.9, 'lr': scale_lr * 1e-1},
                {'epoch': 30, 'lr': scale_lr * 1e-2},
                {'epoch': 60, 'lr': scale_lr * 1e-3},
                {'epoch': 80, 'lr': scale_lr * 1e-4}
            ]
            self.data_regime = [
                {'epoch': 0, 'input_size': 128, 'batch_size': 512},
                {'epoch': 80, 'input_size': 224, 'batch_size': 128},
            ]
            self.data_eval_regime = [
                {'epoch': 0, 'input_size': 128, 'batch_size': 1024},
                {'epoch': 80, 'input_size': 224, 'batch_size': 512},
            ]
        else:
            self.regime = [
                {'epoch': 0, 'optimizer': 'SGD', 'lr': 1e-1, 'momentum': 0.9},
                {'epoch': 30, 'lr': 1e-2},
                {'epoch': 60, 'lr': 1e-3},
                {'epoch': 80, 'lr': 1e-4}
            ]
```
I believe I have also trained the model ( I used this [implementation](https://github.com/marvis/pytorch-mobilenet/blob/master/main.py) based on the same preprocessing and initial learning rate and weight decay but managed to only get 64~65% top1. 
Would you mind explaining a bit more whats different here? 
Thanks alot in advance

**Comment by fmassa** [comment]
We have added MobileNetV2 in master with pre-trained weights, so this might not be necessary anymore for now. Closing it, let me know if you disagree
